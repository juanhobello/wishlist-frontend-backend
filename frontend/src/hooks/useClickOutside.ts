import { useEffect } from "react";
import type { RefObject } from "react";

export default function useClickOutside(
  elementRef: RefObject<HTMLElement>,
  callback: VoidFunction
) {
  useEffect(() => {
    const handleClickOutside = (event: Event) => {
      if (
        elementRef.current &&
        !elementRef.current.contains(event.target as Node)
      ) {
        callback();
      }
    };

    document.addEventListener("click", handleClickOutside, true);
    return () => {
      document.removeEventListener("click", handleClickOutside, true);
    };
  }, [elementRef, callback]);
}
