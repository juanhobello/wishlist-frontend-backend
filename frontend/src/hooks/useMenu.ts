import { useState } from "react";

export const useMenu = () => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = !!anchorEl;

  const handleOpen = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return { anchorEl, open, handleOpen, handleClose };
};
  