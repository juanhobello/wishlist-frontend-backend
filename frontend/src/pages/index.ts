import { Home } from "./Home";
import { Wishlist } from "./Wishlist";
import { NotFound } from "./NotFound";

export { Home, Wishlist, NotFound };
