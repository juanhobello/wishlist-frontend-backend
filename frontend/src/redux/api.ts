import {
  BaseQueryFn,
  createApi,
} from "@reduxjs/toolkit/query/react";
import axios, { AxiosRequestConfig, Method } from "axios";

interface AxiosBaseQueryArgs {
  url: string;
  method: Method;
  data?: AxiosRequestConfig["data"];
  params?: AxiosRequestConfig["params"];
}

interface AxiosBaseQueryError {
  data?: unknown;
  status?: number;
}

const axiosBaseQuery =
  (): BaseQueryFn<AxiosBaseQueryArgs, unknown, AxiosBaseQueryError> =>
  async ({ url, method, data, params }, api) => {
    const serverUrl = `http://${process.env.VITE_SERVE_URL}/${url}`;

    try {
      const result = await axios({
        url: serverUrl,
        method,
        data,
        params,
        timeout: 30000,
        cancelToken: new axios.CancelToken((executor) => {
          api.signal.addEventListener("abort", () => executor());
        }),
      });

      return { data: result.data };
    } catch (error) {
      if (axios.isAxiosError(error)) {
        return {
          error: { data: error.response?.data, status: error.response?.status },
        };
      }
      
      if (error instanceof Error) {
        return { error: { data: error.message, status: undefined } };
      }
      return { error: { data: "Unknown error", status: undefined } };
    }
  };

export const api = createApi({
  baseQuery: axiosBaseQuery(),
  tagTypes: ["Products"],
  endpoints: () => ({}),
});
