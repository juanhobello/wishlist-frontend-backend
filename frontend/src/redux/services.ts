export * from "./modules/products/service";