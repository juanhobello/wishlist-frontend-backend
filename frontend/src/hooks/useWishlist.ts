import { useAppDispatch, useAppSelector } from "@/redux";
import { addProductWishlist, removeProductWishlist } from "@/redux/slices";
import { Product } from "@/redux/types";
import { useMemo } from "react";

export const useWishlist = () => {
  const dispatch = useAppDispatch();
  const wishlist = useAppSelector((state) => state.wishlist);
  const hasWishlistItems = wishlist.length > 0;

  const wishlistCodes = useMemo(
    () => new Set(wishlist.map((product) => product.code)),
    [wishlist]
  );

  const isItemInWishlist = (code: string) => wishlistCodes.has(code);

  const addItemWishlist = (product: Product) => {
    dispatch(addProductWishlist(product));
  };

  const removeItemWishlist = (code: string) => {
    dispatch(removeProductWishlist(code));
  };

  const toggleWishlistItem = (product: Product) => {
    if (isItemInWishlist(product.code)) {
      removeItemWishlist(product.code);
    } else {
      addItemWishlist(product);
    }
  };

  return {
    wishlist,
    isItemInWishlist,
    hasWishlistItems,
    toggleWishlistItem,
    removeItemWishlist,
  };
};
