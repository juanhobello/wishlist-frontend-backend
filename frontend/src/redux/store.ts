import { api } from "./api";
import { combineReducers, Reducer, Action, Store } from "redux";
import { configureStore, ThunkDispatch } from "@reduxjs/toolkit";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import autoMergeLevel2 from "redux-persist/lib/stateReconciler/autoMergeLevel2";
import productReducer, { productInitialState } from "./modules/products/slice";
import wishlistReducer, { wishlistInitialState } from "./modules/wishlist/slice";

type ApiState = ReturnType<typeof api.reducer>;

export interface AppState {
  products: typeof productInitialState;
  wishlist: typeof wishlistInitialState;
  api: ApiState;
}

const rootReducer: Reducer<AppState, Action> = combineReducers({
  products: productReducer,
  wishlist: wishlistReducer,
  [api.reducerPath]: api.reducer,
});

const persistConfig = {
  key: "root",
  storage,
  version: 0,
  stateReconciler: autoMergeLevel2,
  blacklist: [api.reducerPath],
};

const persistedReducer = persistReducer<AppState>(persistConfig, rootReducer);

export const store: Store<AppState> = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: ["persist/PERSIST"],
        ignoredPaths: ["register"],
      },
    }).concat(api.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = ThunkDispatch<RootState, unknown, Action> &
  typeof store.dispatch;
