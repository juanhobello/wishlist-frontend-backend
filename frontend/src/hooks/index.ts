import useClickOutside from "./useClickOutside";
import { useWishlist } from "./useWishlist";

export { useClickOutside, useWishlist };