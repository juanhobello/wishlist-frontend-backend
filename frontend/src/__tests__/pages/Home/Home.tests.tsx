import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { Home } from '@/pages';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { store } from '@/redux/store';
import { useGetProductsQuery } from '@/redux/services';
import { useWishlist } from '@/hooks';
import { products } from '@/__mocks__/producst.mock';

jest.mock('@/redux/services', () => ({
  useGetProductsQuery: jest.fn(),
}));

jest.mock('@/hooks', () => ({
  useWishlist: jest.fn(),
}));

const localStorageMock: unknown = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  clear: jest.fn(),
};

global['localStorage'] = localStorageMock as Storage;

const setupHomePage = () => (
  <Provider store={store}>
    <BrowserRouter>
      <Home />
    </BrowserRouter>
  </Provider>
)

describe('Home Page', () => {
  beforeEach(() => {
    localStorage.clear();
  });

  it('matches snapshot', () => {
    (useGetProductsQuery as jest.Mock).mockReturnValue({
      data: products,
      isLoading: false,
      isError: false,
    });

    (useWishlist as jest.Mock).mockReturnValue({
      isItemInWishlist: jest.fn(() => false),
      toggleWishlistItem: jest.fn(),
    });

    const { container } = render(setupHomePage());

    expect(container).toMatchSnapshot();
  });

  it('should render the product correctly', () => {
    (useGetProductsQuery as jest.Mock).mockReturnValue({
      data: products,
      isLoading: false,
      isError: false,
    });

    (useWishlist as jest.Mock).mockReturnValue({
      isItemInWishlist: jest.fn(() => false),
      toggleWishlistItem: jest.fn(),
    });

    render(setupHomePage());

    expect(screen.getByText(products[0].name)).toBeInTheDocument();
    const likeButton = screen.getByLabelText(`like-button-${products[0].code}`);
    expect(likeButton).toBeInTheDocument();
  });

  it('should add a product to the wishlist and store it in localStorage', () => {
    const toggleWishlistItem = jest.fn((product) => {
      const wishlist = JSON.parse(localStorage.getItem('persist:root') || '{}').wishlist || [];
      localStorage.setItem(
        'persist:root',
        JSON.stringify({ wishlist: [...wishlist, product] })
      );
    });

    (useGetProductsQuery as jest.Mock).mockReturnValue({
      data: products,
      isLoading: false,
      isError: false,
    });

    (useWishlist as jest.Mock).mockReturnValue({
      isItemInWishlist: jest.fn(() => false),
      toggleWishlistItem,
    });

    render(setupHomePage());

    const likeButton = screen.getByLabelText(`like-button-${products[0].code}`);
    fireEvent.click(likeButton);

    const storedState = JSON.parse(localStorage.getItem('persist:root') || '{}');
    const storedWishlist = storedState.wishlist;
    expect(storedWishlist).toHaveLength(1);
    expect(storedWishlist[0]).toEqual(products[0]);
  });
});
