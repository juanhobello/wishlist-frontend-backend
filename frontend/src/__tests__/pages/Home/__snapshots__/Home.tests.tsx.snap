// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page matches snapshot 1`] = `
<div>
  <ol
    class="breadcrumbs-list"
  >
    <li>
      <a
        class="breadcrumbs-link"
        data-discover="true"
        href="/"
      >
        Home
      </a>
    </li>
  </ol>
  <hr
    class="divider"
  />
  <div
    class="products-list"
  >
    <div
      class="product-card"
    >
      <span
        class="float-button"
      >
        <button
          aria-label="like-button-D22-2077-006"
          class="heart-button"
        >
          <svg
            fill="none"
            font-size="20"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
            />
          </svg>
        </button>
      </span>
      <img
        alt="Tênis Nike Revolution 7 Feminino - Preto+Branco"
        class="img"
        src="https://static.netshoes.com.br/produtos/tenis-nike-revolution-7-feminino/26/JD8-6363-026/JD8-6363-026_zoom1.jpg?ts=1708096757?ims=544x"
      />
      <div
        class="title-product"
      >
        <p>
          Tênis Nike Revolution 7 Feminino - Preto+Branco
        </p>
      </div>
      <div
        class="rating"
      >
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-50"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="50%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="50%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-50)"
          />
        </svg>
        <p>
          4.5
        </p>
      </div>
      <s
        class="text"
      >
        R$ 399,99
      </s>
      <span
        class="price"
      >
        R$ 303,99
      </span>
    </div>
    <div
      class="product-card"
    >
      <span
        class="float-button"
      >
        <button
          aria-label="like-button-NQQ-4378-028"
          class="heart-button"
        >
          <svg
            fill="none"
            font-size="20"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
            />
          </svg>
        </button>
      </span>
      <img
        alt="Bola de Futebol Society Penalty 8 X - Branco+Verde Limão"
        class="img"
        src="https://static.netshoes.com.br/produtos/bola-de-futebol-society-penalty-8-x/78/D23-5694-078/D23-5694-078_zoom1.jpg?ts=1695700132?ims=544x"
      />
      <div
        class="title-product"
      >
        <p>
          Bola de Futebol Society Penalty 8 X - Branco+Verde Limão
        </p>
      </div>
      <div
        class="rating"
      >
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-100"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-100)"
          />
        </svg>
        <svg
          aria-label="Star"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="grad-0"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="0%"
            >
              <stop
                offset="0%"
                style="stop-opacity: 1;"
              />
              <stop
                offset="0%"
                style="stop-opacity: 1;"
              />
            </lineargradient>
          </defs>
          <path
            d="M8 .192l1.889 5.812h6.111l-4.944 3.594 1.889 5.813-4.944-3.594-4.944 3.594 1.889-5.813-4.944-3.594h6.111z"
            fill="url(#grad-0)"
          />
        </svg>
        <p>
          4.0
        </p>
      </div>
      <s
        class="text"
      >
        R$ 220,00
      </s>
      <span
        class="price"
      >
        R$ 174,99
      </span>
    </div>
  </div>
</div>
`;
